Fix local commodity definitions affecting data passed to EDDN

Tkael opened the pull request:
Resolves #1906, resolves #1931, & resolves #1953.
- Moves parsing CAPI `profile` results from `CompanionAppService.cs` into `CompanionAppProfileFromJson.cs` (to separate authentication from parsing the results)
- Sorts commodities by EliteId and updates commodity definitions to provide a short term resolution for #1906.
  - SanumaMEAT was SanumaDecorativeMeat
  - Wolf1301Fesh was WolfFesh
  - CD75CatCoffee was CD-75KittenBrandCoffee
  - AerialEdenApple was EdenApplesOfAerial
  - UzumokuLowGWings was UzumokuLow-GWings
- Passes only original data to the EDDN responder.
- Updates profile.cs to also only pass original data
- Removes the EDID property from the `FrontierApiCommander.cs` class (the ID returned from the Frontier API was found to not match the FID returned from the player journal)
- Implements an optional enum for commodity stock and demand brackets, standardizing the data type for brackets (which were either `dynamic` or `int` for the `CommodityMarketQuote` or `MarketInfo` classes respectively)
- Uses a common `MarketInfoItem` class to capture commodity market quote information from the Frontier API or from market.json, ready to pass to the EDDN responder, and uses the `CommodityMarketQuote` class when we are ready to embed an EDDI data definition.
- Exposes imports, exports, and prohibited commodities as localized commodity objects and updates documentation accordingly. Updates the `Cargo report` script (to use commodity objects for the station `prohibited` list)
- Revised the `Market information updated` script to resolve #1953.

Tkael committed:
Sort commodities by EliteId and update commodity definitions to provide a short term resolution for #1906.
- SanumaMEAT was SanumaDecorativeMeat
- Wolf1301Fesh was WolfFesh
- CD75CatCoffee was CD-75KittenBrandCoffee
- AerialEdenApple was EdenApplesOfAerial
- UzumokuLowGWings was UzumokuLow-GWings

Tkael committed:
Pass only original data to the EDDN responder.
- Split parsing of the cAPI profile to a seperate class
- Update profile.cs to only pass original data
- Remove the EDID property from the `FrontierApiCommander.cs` class (the ID returned from the Frontier API was found to not match the FID returned from the player journal)
- Use an optional enum for commodity stock and demand brackets
- Use the `eddnCommodityMarketQuote` class to capture commodity market quote information ready to pass to EDDN.
- Use the `CommodityMarketQuote` class to embed EDDI definitions

Tkael committed:
Make imports, exports, and prohibited commodities accessible with localized definitions.

Tkael committed:
Update the `Cargo report` script to use commodity objects for the station prohibited commodities list.

GioVAX commented on CompanionAppService/CompanionAppProfileFromJson.cs:
```diff
@@ -0,0 +1,297 @@
+        /// <summary>Create a  profile given the results from a /profile call</summary>
+        public static Profile ProfileFromJson(string data, DateTime timestamp)
+        {
+            Profile profile = null;
```
This temp variable is not necessary and could be removed, resulting in simpler code.

Tkael replied:
Agreed. Update pushed.

GioVAX commented on CompanionAppService/CompanionAppProfileFromJson.cs:
```diff
@@ -0,0 +1,297 @@
+                timestamp = timestamp
+            };
+
+            if (json["commander"] != null)
```
I know we want to change the way we parse JSON in the future, but this is an opportunity to try a new approach and see how it works.
Also, it will reduce the number of changes we'll need to do once we decide to migrate ;)

Tkael replied:
Changing the overall parsing strategy adds additional risk to this PR. I don't think that it makes sense for us to do that right now.

Tkael replied:
Additionally, you aren't being very clear about the method that you think we should be adopting here. ;-)

GioVAX replied:
That's true 😆 
I was thinking to Newtonsoft `JsonConvert.DeserializeObject<T>(json)`

GioVAX commented on CompanionAppService/CompanionAppProfileFromJson.cs:
```diff
@@ -0,0 +1,297 @@
+        // Obtain the list of station economies from the profile
+        public static List<ProfileEconomyShare> EconomiesFromProfile(dynamic json)
+        {
+            List<ProfileEconomyShare> Economies = new List<ProfileEconomyShare>();
```
Should this be `economies`?
Also, in other parts of the code in the declaration, we use `var`.

Tkael replied:
I'll revise to `economyShares`.

GioVAX commented on CompanionAppService/CompanionAppProfileFromJson.cs:
```diff
@@ -0,0 +1,297 @@
+            {
+                // shipyard_list is a JObject containing JObjects but let's code defensively because FDev
+                var shipyardList = json["lastStarport"]["ships"]["shipyard_list"].Children();
+                foreach (JToken shipToken in shipyardList.Values())
```
Any reason not to use LINQ?

```
var edShipyardShips = shipyardList.Values()
    .Select( v => ShipyardShipFromProfile( v as JObject );
```

Tkael replied:
Looks like `CapiShipyardTests` validates changes to how we parse this so this is a low risk change. Ok.

GioVAX commented on DataDefinitions/CommodityDefinition.cs:
```diff
@@ -519,6 +525,38 @@ private static bool ignoredCommodity(string name)
             }
         }
 
+        private static bool obsoleteCommodityEdName(string name, out string correctedName)
```
We could remove the bool, and just return the normalized commodity name (changed if necessary, or as is otherwise).
Who cares if the name was actually changed?

Tkael replied:
Ok

GioVAX commented on DataDefinitions/EddnCommodityMarketQuote.cs:
```diff
@@ -0,0 +1,135 @@
+            demandBracket = capiJSON["demandBracket"]?.ToObject<CommodityBracket?>();
+            if (capiJSON["statusFlags"] != null)
+            {
+                foreach (string statusFlag in capiJSON["statusFlags"])
```
If capiJSON["statusFlags"] is a `Enumerable`, this could become
`statusFlags = capiJSON["statusFlags"].Cast<string>();`

Tkael replied:
As far as I'm aware, statusFlags is not an Enumberable.

Tkael replied:
I'm getting an invalid cast exception with the proposed code.
Note: `statusFlags` is a list so I appended `.ToList()`

GioVAX replied:
Ok.

GioVAX commented on DataDefinitions/MarketInfo.cs:
```diff
@@ -23,13 +23,13 @@ public class MarketInfo
         public int stock { get; set; }
         [JsonProperty]
-        public int stockbracket { get; set; }
+        public CommodityBracket? stockbracket { get; set; }
```
Just curious: why optional?
Previously `int` was not optional.

Tkael replied:
We've seen from the Frontier API that the enum can be null. Making it optional allows for that possibility in Market.json

GioVAX replied:
Thanks for the explanation 😃

GioVAX commented on EDDI/EDDI.cs:
```diff
@@ -1572,42 +1572,50 @@ private bool eventLiftoff(LiftoffEvent theEvent)
         {
-            if (allowMarketUpdate)
+            // Don't proceed if we've already viewed the market while docked or when loading pre-existing logs
+            if (allowMarketUpdate && !theEvent.fromLoad)
```
To simplify a bit, we could turn this `if` and return immediately.
E.g.
```
if( theEvent.fromLoad || ! allowMarketUpdate )
    return false;
....
```

Tkael replied:
ok

GioVAX commented on EDDI/EDDI.cs:
```diff
@@ -1616,14 +1624,19 @@ private bool eventMarket(MarketEvent theEvent)
         {
-            if (allowOutfittingUpdate)
+            // Don't proceed if we've already viewed outfitting while docked or when loading pre-existing logs
+            if (allowOutfittingUpdate && !theEvent.fromLoad)
```
Same as above.

Tkael replied:
ok

GioVAX commented on EDDI/EDDI.cs:
```diff
@@ -1616,14 +1624,19 @@ private bool eventMarket(MarketEvent theEvent)
+
+                    // Update the current station outfitting
                     List<EddiDataDefinitions.Module> modules = new List<EddiDataDefinitions.Module>();
                     foreach (OutfittingInfo item in info.Items)
```
This block could be simplified with LINQ.
```
var modules = info.Items
    .Select( i => EddiDataDefinitions.Module.FromOutfittingInfo(i))
    .Where( i => i ! null )
    .ToList();
```

Tkael replied:
ok

GioVAX commented on EDDI/EDDI.cs:
```diff
@@ -1661,7 +1667,8 @@ private bool eventOutfitting(OutfittingEvent theEvent)
         {
-            if (allowShipyardUpdate)
+            // Don't proceed if we've already viewed outfitting while docked or when loading pre-existing logs
+            if (allowShipyardUpdate && !theEvent.fromLoad)
```
As previous `if`s

Tkael replied:
ok

GioVAX commented on EDDI/EDDI.cs:
```diff
@@ -1678,12 +1685,14 @@ private bool eventShipyard(ShipyardEvent theEvent)
                             ships.Add(ship);
```
This block could be simplified as above.

Tkael replied:
ok

Tkael committed:
Simplify the ProfileFromJson() method (as suggested by @Giovax)

Tkael committed:
Prefer the more descriptive (and more conformant to standards) variable name `economyShares` over `Economies`

Tkael committed:
Replace `obsoleteCommodityEdName()` method with more direct `correctedCommodityEdName()` method.

Tkael committed:
Simplify converting outfitting and shipyard items to objects with Linq.

Tkael committed:
Invert ifs to reduce nesting in our handlers for Market, Outfitting, and Shipyard events.

Tkael committed:
Simplify shipyard parsing using Linq.

Tkael committed:
Add a status flag to our commodity unit tests.

Tkael committed:
Move info reader content into the base event

(so that we can more easily inject content for unit testing the events)

Tkael committed:
Rename classes for clarity

Tkael committed:
Pass "Rare" as a status flag, use lowercase names for better string replacement.

Tkael committed:
Allow commodities with a meanPrice of 0 (like from carriers) to be passed to EDDN.

Tkael committed:
Consolidate similar classes EDDNCommodityMarketQuote and MarketInfoItem into one.

Tkael commented:
Updated to resolve #1931, improve testing, and consolidate some similar classes.

Tkael committed:
Update commodity definitions (average pricing and EDDB IDs) from https://eddb.io/archive/v6/commodities.json.

Tkael committed:
Consolidate similar classes for outfitting and shipyard items.

Tkael committed:
Enhance cAPI parsing testing.

Tkael committed:
Add a unit test for the deserialized output of market.json

Tkael committed:
Refactor and cover ProfileStation()

Tkael committed:
Test the `UpdateStation()` method in the Profile class.

Tkael committed:
Add a unit test to validate the commodity data sent to EDDN.
Fix category being serialized when it shouldn't.
Fix commodity brackets being reported as null rather than as an empty string.

Tkael committed:
clean up test vars

Tkael committed:
Fix test bug w/ deserialized average pricing.

Tkael committed:
Compare commodity ednames with `ToLowerInvariant()`.

Tkael committed:
Resolve #1953.

Tkael committed:
Permit journal-based updates whenever the profile returns no data for an endpoint.
Allows journal based /shipyard EDDN updates when the cAPI profile returns no data for an endpoint (like reported in https://issues.frontierstore.net/issue-detail/18441).

Tkael committed:
Add a definition test for reading market.json. Enhance the definition test for a commodity from CAPI.
